File: src/MappingEdu.Core.Domain/CustomDetailMetadata.cs
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

using System;
using System.Collections.Generic;
using System.Linq;
using MappingEdu.Core.Domain.System;

namespace MappingEdu.Core.Domain
{
    public class CustomDetailMetadata : Entity, ICloneable<CustomDetailMetadata>
    {
        public Guid CustomDetailMetadataId { get; set; }

        public string DisplayName { get; set; }

        protected override Guid Id
        {
            get { return CustomDetailMetadataId; }
        }

        public bool IsBoolean { get; set; }

        public bool IsCoreDetail { get; set; }

        public virtual MappedSystem MappedSystem { get; set; }

        public Guid MappedSystemId { get; set; }

        public virtual ICollection<SystemItemCustomDetail> SystemItemCustomDetails { get; set; }

        public CustomDetailMetadata()
        {
            SystemItemCustomDetails = new HashSet<SystemItemCustomDetail>();
        }

        public CustomDetailMetadata Clone()
        {
            var clone = new CustomDetailMetadata()
            {
                CustomDetailMetadataId = CustomDetailMetadataId,
                DisplayName = DisplayName,
                IsBoolean = IsBoolean,
                IsCoreDetail = IsCoreDetail,
            };

            return clone;
        }
    }
}

File: src/MappingEdu.Web.UI/Client/App/Repositories/CustomDetailMetadata.ts
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.


// ****************************************************************************
// Module app.repositories.custom-detail-metadata
//

var m = angular.module('app.repositories.custom-detail-metadata', ['restangular']);


// ****************************************************************************
// Interface CustomDetailMetadata
//
 
interface ICustomDetailMetadata extends restangular.IElement {
    DisplayName: string,
    IsBoolean: boolean,
    IsCore: boolean,
    CustomDetailMetadataId: string,
    DataStandardId: string
}


// ****************************************************************************
// Interface ICustomDetailMetadataRepository
//

interface ICustomDetailMetadataRepository {
    getAllByDataStandard(dataStandardId: string): angular.IPromise<Array<ICustomDetailMetadata>>;
    create(dataStandardId: string, customDetail: any): angular.IPromise<ICustomDetailMetadata>;
    save(dataStandardId: string, customDetail: any): angular.IPromise<ICustomDetailMetadata>;
    remove(dataStandardId: string, customDetailMetadataId: string): angular.IPromise<void>;
}

// ****************************************************************************
// SystemConstants repository
//

m.factory('app.repositories.custom-detail-metadata', ['Restangular', (restangular: restangular.IService) => {

    // extend 

    restangular.extendModel('CustomDetailMetadata', (model: ICustomDetailMetadata) => {
        return model;
    });

    // methods

    function getAllByDataStandard(dataStandardId: string) {
        return restangular.one('CustomDetailMetadata', dataStandardId).get();
    }

    function create(dataStandardId: string, customDetail: any) {
        return restangular.one('CustomDetailMetadata', dataStandardId).customPOST(customDetail);
    }

    function save(dataStandardId: string, customDetail: any) {
        return restangular.one('CustomDetailMetadata', customDetail.CustomDetailMetadataId).one('DataStandard', dataStandardId).customPUT(customDetail);
    }

    function remove(dataStandardId: string, customDetailMetadataId: string) {
        return restangular.one('CustomDetailMetadata', customDetailMetadataId).one('DataStandard', dataStandardId).remove();
    }

    var repository: ICustomDetailMetadataRepository = {
        getAllByDataStandard: getAllByDataStandard,
        create: create,
        save: save, 
        remove: remove
    };

    return repository;
}]); 